### PR TITLE
Command scopes, better semantics of commands, different /send and /request for groups and personal chats

### PR DESCRIPTION
**Pull request by alwx**
Ok, this PR is quite big, and there will be a lot of text here.

## What happens?

This PR is a part of API refactoring/restructuring, and this PR changes the way commands look and work. It removes any need in hardcoded `/send` and `/request`-related things and also makes mailman-like behavior accessible for 3rd-party developers. It also introduces scopes for commands and removes 99% of hardcoded things. By doing it, we make our API better and remove almost all the internal "magic" which was previously unavailable for 3rd-party developers. 

### 1. Scopes

The most important part of this PR is scopes for commands.
The best way to demonstrate scopes is this extract from the new `/send`:

```js
var send = {
    name: "send",
    scope: {
        isGlobal: false,
        personalChats: true,
        groupChats: false,
        canUseForDApps: false
    },
    icon: "money_white",
    color: "#5fc48d",
    title: I18n.t('send_title'),
    description: I18n.t('send_description'),
    params: paramsSend,
    validator: validateSend,
    handler: handleSend,
    preview: previewSend,
    shortPreview: shortPreviewSend
};
```

The important thing is that we can create two or more commands with the same name, but different scopes and a different set of parameters. It allows us to get rid of contact selector for `/send` in 1-to-1 chats and allows 3rd-party devs to make commands more flexible.

### 2. Mailman

Mailman is no longer something special. Everybody can create the same contact whos commands will be mixed with other contact's commands. In combination with scopes, it allows creating something super crazy and flexible.

### 3. Less magic for /send and /request

There are still some hardcoded things for /send and /request, but the number of them has decreased drastically.

### 4. Commands model

I moved some common commands-related things to a separate model. It is less complicated and structured better than it was before.

### 5. Tons of code quality improvements

Yay!

### 6. One more thing — text input should work faster

We don't call `:update-suggestions` each time user types a character since now, so the speed of text input should increase.

### CODE REVIEW

File: src/status_im/chat/models/commands.cljs [hunk 1]
```diff
@@ -0,0 +1,124 @@
+(ns status-im.chat.models.commands
+  (:require [status-im.chat.constants :as chat-consts]
+            [status-im.bots.constants :as bots-constants]
+            [clojure.string :as str]
+            [taoensso.timbre :as log]))
+
+(defn scope->int [{:keys [global? registered-only? personal-chats? group-chats? can-use-for-dapps?]}]
```
Comment (janherich) [inline src/status_im/chat/models/commands.cljs]: Could we rename this to `scope->hash` and provide proper docstring describing how hash is computed by bitwise OR operation ?
Otherwise it could be quite cryptic at first glance if the person working on the code is not familiar with how catalog of commands is structured in `status.js` and how scope hash is used as second level index.

**Review by janherich — APPROVED** [review]
I think the scopes are step in the right direction, I like the, also how it simplifies suggestions on the RN side.
The only thing which I'm little bit uneasy with is the duplication in `transactor_group/bot.js` and `transactor_personal/bot.js` commands, we should lift common code into some other file up in the hierarchy, so it's reused.

**Comment by alwx** [comment]
@janherich also don't like this kind of duplication (`transactor_group` and `transactor_personal`). However, if I combine them into one file, this file will be incredibly big. I think the best way we can solve this problem is by introducing the ability to specify entry points and have several files like `bot.js`.
It is not related to this task, but I will think about a better and not time-consuming way of solving this problem.

File: src/status_im/chat/models/commands.cljs (hunk 1 above)
Comment (alwx) [reply src/status_im/chat/models/commands.cljs]: This scope->int calculation is pretty self-descriptive, isn't it? And doing this kind of calculation is a common practice.

Comment (janherich) [reply src/status_im/chat/models/commands.cljs]: Well, when somebody looks at any place in code where it's used (mostly when constructing path for jail call), `scope->int` is little bit vague and she/he needs to jump to function definition before it starts making sense, but with `scope->hash` it's immediately clear that we are computing hash out of the scope map.
But not a big deal, I'm fine with either way.

File: .re-natal [hunk 2]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "StatusIm",
   "interface": "reagent",
-  "androidHost": "localhost",
+  "androidHost": "10.0.3.2",
```
Comment (oskarth) [inline .re-natal:4]: Can you remove this from the PR?

File: env/dev/env/android/main.cljs [hunk 3]
```diff
@@ -15,8 +15,8 @@
   (re-frame.core/dispatch [:load-commands!]))
 
 (figwheel/watch-and-reload
-  :websocket-url "ws://localhost:3449/figwheel-ws"
+  :websocket-url "ws://10.0.3.2:3449/figwheel-ws"
```
Comment (oskarth) [inline env/dev/env/android/main.cljs:18]: Same here with entire file (and `env/ios`)

File: ios/Podfile.lock [hunk 4]
```diff
@@ -44,4 +44,4 @@ SPEC CHECKSUMS:
 
 PODFILE CHECKSUM: d05403e3fce258b6985bd3b614ffe02a0e01b9f7
 
-COCOAPODS: 1.3.1
+COCOAPODS: 1.2.1
```
Comment (oskarth) [inline ios/Podfile.lock]: Please don't downgrade COCOAPODS version

File: ios/StatusIm.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -947,6 +947,7 @@
 				A97BA941B2FB44B4B66EE6D3 /* Frameworks */,
 				1E7837547A9A40E18AD63CF3 /* Resources */,
 				5C1C8762251D6EF495FB2384 /* Pods */,
+				B280E73E1F7E6CB000B5E1A6 /* Recovered References */,
```
Comment (oskarth) [inline ios/StatusIm.xcodeproj/project.pbxproj]: What changes does API have to do with iOS project settings?

File: resources/js/bots/transactor_group/bot.js [hunk 6]
```diff
@@ -284,7 +284,11 @@ function validateSend(params, context) {
         params["bot-db"] = {};
     }
 
-    if (!params["bot-db"]["public"] || !params["bot-db"]["public"]["recipient"] || !params["bot-db"]["public"]["recipient"]["address"]) {
+    console.log(context);
```
Comment (oskarth) [inline resources/js/bots/transactor_group/bot.js]: Does this log have sensitive info? Do we need it?

File: resources/js/status.js [hunk 7]
```diff
@@ -6,13 +6,21 @@ var _status_catalog = {
     },
     status = {};
 
+function transformScope(scope) {
```
Comment (oskarth) [inline resources/js/status.js]: A comment on what this does would be great, looks like a bitmask of some kind?

File: src/status_im/chat/handlers/send_message.cljs [hunk 8]
```diff
@@ -134,7 +134,7 @@
   (u/side-effect!
     (fn [_ [_ {:keys [command chat-id]}]]
       (let [{:keys [to-message]} command]
-        (when to-message
+        (when (and to-message (not= to-message :any))
```
Comment (oskarth) [inline src/status_im/chat/handlers/send_message.cljs]: What does this do in plain english? Don't quite understand

File: src/status_im/chat/models/commands.cljs [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+  (let [global-commands (get-possible-global-commands db)
+        commands        (get-possible-commands db)
+        account         (get accounts current-account-id)
+        commands        (-> (into [] global-commands)
+                            (into commands))
+        {chat-contacts :contacts} (get chats chat-id)]
+    (->> commands
+         (remove (fn [{:keys [scope]}]
+                   (or
+                     (and (:registered-only? scope)
```
Comment (oskarth) [inline src/status_im/chat/models/commands.cljs]: Can we bring out these predicates into fns in let? So it is more clear semantically what conditions must be met. Something like:

```
(or (personal-chat? scope) foo? bar? ...)
```

(Can bind `chat-contacts` with partial fn, for example)

File: src/status_im/chat/models/commands.cljs [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+                     (and (not (:personal-chats? scope))
+                          (= (count chat-contacts) 1))
+                     (and (not (:group-chats? scope))
+                          (> (count chat-contacts) 1))
+                     (and (not (:can-use-for-dapps? scope))
+                          (every? (fn [{:keys [identity]}]
+                                    (get-in contacts [identity :dapp?]))
+                                  chat-contacts))))))))
+
+(defn parse-command-message-content
```
Comment (oskarth) [inline src/status_im/chat/models/commands.cljs]: Since this is in commands namespace, do we need command in name?

File: src/status_im/chat/views/message/request_message.cljs [hunk 11]
```diff
@@ -70,35 +71,35 @@
...
+                               command)
+            on-press-handler (if (:execute-immediately? command)
+                               #(dispatch [:execute-command-immediately command])
+                               (when (and (not answered?) status-initialized?)
+                                 #(set-chat-command message-id command)))]
         [view st/comand-request-view
          [touchable-highlight
           {:on-press on-press-handler}
           [view st/command-request-message-view
-           (if (and @markup
```
Comment (oskarth) [inline src/status_im/chat/views/message/request_message.cljs]: Huh, lots of weird derefs in existing code here. I think this might get rid of some bugs, nice one :)

File: src/status_im/data_store/realm/schemas/account/core.cljs [hunk 12]
```diff
@@ -60,4 +62,10 @@
               {:schema        v14/schema
                :schemaVersion 14
                :migration     v14/migration}
+              {:schema        v15/schema
+               :schemaVersion 15
+               :migration     v15/migration}
+              {:schema        v16/schema
+               :schemaVersion 16
```
Comment (oskarth) [inline src/status_im/data_store/realm/schemas/account/core.cljs]: Do we need two schema bumps? What is the reason for not doing it in v15 alone?

File: src/status_im/ui/screens/contacts/subs.cljs [hunk 13]
```diff
@@ -26,9 +26,10 @@
   :all-added-contacts
   :<- [:get-contacts]
   (fn [contacts]
-    (->> (remove (fn [[_ {:keys [pending? whisper-identity]}]]
+    (->> contacts
+         (remove (fn [[_ {:keys [pending? mixable? whisper-identity]}]]
                    (or (true? pending?)
-                       (bots-constants/hidden-bots whisper-identity))) contacts)
+                       (true? mixable?))))
```
Comment (oskarth) [inline src/status_im/ui/screens/contacts/subs.cljs]: Don't think you need `(true? ...)` here, as long as `mixable` is truthy or falsy, which it should be based on its name.

File: test/cljs/status_im/test/chat/models/input.cljs [hunk 14]
```diff
@@ -3,7 +3,7 @@
             [status-im.chat.models.input :as input]))
 
 (def fake-db
-  {:global-commands   {:command1 {:name "global-command1"}}
+  {:global-commands   {:command1 '({:name "global-command1"})}
```
Comment (oskarth) [inline test/cljs/status_im/test/chat/models/input.cljs]: Can you explain why we need this quoting? It's only tests but I don't understand why

**Comment by rasom** [comment]
I would prefer to have something like
```js
scopes: ['personal-chat'],
```
instead of
```js
scope: {
        isGlobal: false,
        personalChats: true,
        groupChats: false,
        canUseForDApps: false
    },
```
just a suggestion

File: test/cljs/status_im/test/chat/models/input.cljs (hunk 14 above)
Comment (rasom) [reply test/cljs/status_im/test/chat/models/input.cljs]: just to create a list i suppose. But maybe better to use `list` here

Comment (oskarth) [reply test/cljs/status_im/test/chat/models/input.cljs]: If it is just changing collection type from map to list-like a vector seems more idiomatic. Is there something list-specific about this?

File: src/status_im/data_store/realm/schemas/account/v16/core.cljs [hunk 15]
```diff
@@ -0,0 +1,40 @@
+(ns status-im.data-store.realm.schemas.account.v16.core
```
Comment (rasom) [inline src/status_im/data_store/realm/schemas/account/v16/core.cljs]: @alwx can you merge v15 and v16?

File: src/status_im/data_store/realm/schemas/account/v12/contact.cljs [hunk 16]
```diff
@@ -0,0 +1,47 @@
+(ns status-im.data-store.realm.schemas.account.v12.contact
```
Comment (rasom) [inline src/status_im/data_store/realm/schemas/account/v12/contact.cljs]: Why are we adding v12 contact here?

File: src/status_im/data_store/messages.cljs [hunk 17]
```diff
@@ -113,7 +113,7 @@
       (data-store/save message'))))
 
 (defn update 
-  [{:keys [message-id] :as message}]
+  [{:keys [message-id preview] :as message}]
```
Comment (rasom) [inline src/status_im/data_store/messages.cljs]: `preview` is not used

File: src/status_im/data_store/messages.cljs [hunk 18]
```diff
@@ -5,7 +5,8 @@
             [status-im.utils.utils :refer [update-if-present]]
             [clojure.walk :refer [stringify-keys keywordize-keys]]
             [cljs.reader :refer [read-string]]
-            [status-im.constants :as c])
+            [status-im.constants :as c]
+            [taoensso.timbre :as log])
```
Comment (rasom) [inline src/status_im/data_store/messages.cljs]: was any logging added in this ns?

File: src/status_im/chat/models/commands.cljs (hunk 1 above)
Comment (alwx) [reply src/status_im/chat/models/commands.cljs]: @janherich got it. Will fix.

File: src/status_im/ui/screens/contacts/subs.cljs (hunk 13 above)
Comment (alwx) [reply src/status_im/ui/screens/contacts/subs.cljs]: You're right, it's just a leftover.

**Comment by rasom** [comment]
What does `canUseForDApps` mean?

File: test/cljs/status_im/test/chat/models/input.cljs (hunk 14 above)
Comment (alwx) [reply test/cljs/status_im/test/chat/models/input.cljs]: Let's keep lists here because we use lists in the real db. I will replace `'(...)` with `(list ...)` to improve readability.

File: src/status_im/chat/models/commands.cljs (hunk 9 above)
Comment (alwx) [reply src/status_im/chat/models/commands.cljs]: I tried this approach and must say that it actually makes things a bit more complicated.
Reasons are the following:
1) We also need to pass `account` and `chat-contacts`, or `db` to these functions.
2) If we create separate functions for scopes, they will look this way:

```
(defn registered-only? [scope account]
  (or (:registered-only? scope)
      (:address account)))
```
As for me, the name doesn't really represent what's happening inside + it becomes difficult to understand the function itself.

File: src/status_im/chat/events/input.cljs [hunk 19]
```diff
@@ -136,19 +133,22 @@
...
+                        (commands-model/scope->bit-mask scope)
+                        0)]
                      :params
                      parameter-index
                      :suggestions]
             args    (-> (get-in db [:chats current-chat-id :input-text])
-                        input-model/split-command-args
-                        rest)
+                        (input-model/split-command-args)
+                        (rest))
```
Comment (janherich) [inline src/status_im/chat/events/input.cljs]: Nitpicking, but there is no need to add parenthesis in this case

File: src/status_im/chat/events/input.cljs [hunk 20]
```diff
@@ -283,6 +282,8 @@
                          :chat-id      chat-id
                          :jail-id      (or owner-id jail-id)
                          :content      {:command (:name command)
+                                        :scope   (when-not (:to-message-id metadata)
```
Comment (janherich) [inline src/status_im/chat/events/input.cljs:284]: What's the purpose of this expression, why is :scope nil when `:to-message-id` is present in metadata ?

**Review by janherich — APPROVED** [review]

File: src/status_im/chat/events/input.cljs (hunk 20 above)
Comment (alwx) [reply src/status_im/chat/events/input.cljs:284]: Because we don't use scopes for requests and `:to-message-id` is the indication of request

Comment (janherich) [reply src/status_im/chat/events/input.cljs:284]: So we have no better way of distinguishing command-request message from command message other then inspecting `:to-message-id` key in the metadata ?

**Review by oskarth — APPROVED** [review]

**Comment by annadanchenko** [comment]
#2060 'Property must be of type: string' popup after providing confirmation code [feature/command-scopes] fixed in https://github.com/status-im/status-react/pull/1754/commits/d71e7a3659fd1d5db5e0d3dc6aefbc7ee1dd2d1d
#2061 No Unsigned Transaction screen is shown if send tx in chat with 2 members [feature/command-scopes] This is fixed in `0.9.10-215-g45b94306+` both android and ios
#2072 Icons for /send and /request commands are missing [feature/command-scopes] This is fixed in `0.9.10-215-g45b94306+` both android and ios This is fixed in `0.9.10-215-g45b94306+` both android and ios
#2075 /location command is not retrieving the map [feature/command-scopes] This is fixed in `0.9.10-215-g45b94306+` both android and ios

**Comment by alwx** [comment]
All fixes were merged.

**Comment by jeluard** [comment]
@alwx Can you rebase on develop?
